src/crypto.rs

```diff
@@ -72,6 +122,349 @@ pub fn verify_signed_xml<Bytes: AsRef<[u8]>>(
+    if let Some(object_elem) = object_elem_opt {
+        return Some(object_elem);
+    }
+
+    let sig_info_elem_opt = get_first_child_name_ns(signature_node, "SignedInfo", XMLNS_XML_DSIG);
+    if let Some(sig_info_elem) = sig_info_elem_opt {
+        let ref_elem_opt = get_first_child_name_ns(&sig_info_elem, "Reference", XMLNS_XML_DSIG);
+        if let Some(ref_elem) = ref_elem_opt {
+            if let Some(uri) = ref_elem.get_attribute("URI") {
+                if uri.starts_with('#') {
+                    // prepare a XPointer context
+                    let c_uri = CString::new(&uri[1..]).unwrap();
+                    let ctx_ptr = unsafe {
+                        libxml::bindings::xmlXPtrNewContext(
+                            doc.doc_ptr(),
+                            signature_node.node_ptr(),
+                            std::ptr::null_mut(),
+                        )
+                    };
+                    if ctx_ptr.is_null() {
+                        return None;
+                    }
+                    let ctx = XPathContext { pointer: ctx_ptr };
+
+                    // evaluate the XPointer expression
+                    let obj_ptr = unsafe {
+                        libxml::bindings::xmlXPtrEval(
+                            c_uri.as_ptr() as *const libxml::bindings::xmlChar,
+                            ctx.pointer,
+                        )
+                    };
+                    if obj_ptr.is_null() {
+                        return None;
+                    }
+                    let obj = XPathObject { pointer: obj_ptr };
+
+                    // extract the nodeset from the result
+                    let obj_type = unsafe { (*obj.pointer).type_ };
+                    if obj_type != libxml::bindings::xmlXPathObjectType_XPATH_NODESET {
+                        return None;
+                    }
+                    let obj_nodeset = unsafe { (*obj.pointer).nodesetval };
+                    let nodeset_count = unsafe { (*obj_nodeset).nodeNr };
+
+                    // go through the nodes and find them in the document
+                    for i in 0..nodeset_count {
+                        let node_ptr_ptr = unsafe { (*obj_nodeset).nodeTab.offset(i.try_into().unwrap()) };
+                        let node_ptr = unsafe { *node_ptr_ptr };
+                        if let Some(node) = get_node_by_ptr(&doc.get_root_element().unwrap(), node_ptr) {
+                            return Some(node);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    None
+}
+
+/// Place the signature-verified attributes ([`ATTRIB_SIGVER`] in the given namespace) on the given
+/// element, all its descendants and its whole chain of ancestors (but not necessarily all their
+/// descendants).
+#[cfg(feature = "xmlsec")]
+fn place_signature_verified_attributes(
+    root_elem: libxml::tree::Node,
```

Comment:
Since we seem to be mutating this, is there a reason why we don't annotate this as a mutable borrow? It seems we're abusing the `Rc<Refcell<...>>` and making it harder to track ownership at compile time here.

Reply:
See the comment thread above where we're discussing the `drop(root_elem);` in `place_signature_verified_attributes`.